File: TS/系统/03．技能系统/00．技能模板+函数/02．通用函数/05．召唤物.ts
/** @noSelfInFile */
/**
 * 通用函数 - 召唤物快捷模板
 *
 * 说明：
 * - 配套底层 JASS 源文件：
 *   `JASS/jass复制粘贴/召唤物.j`
 * - 本文件不重写 JASS 召唤逻辑，而是按旧模板要求写入 YDLocal 参数后触发 STES 事件 `OnSummonEvent`
 * - 适合技能侧快速创建/配置召唤物，并继续复用 JASS 端现有字段语义
 */

const jass = require("jass.common") as any;
const jglobals = require("jass.globals") as any;
const STES_Fire_Global = (globalThis as any).STES_Fire as ((self: any, name: string) => void) | undefined;

function 字符串哈希(name: string): number {
  return (jass.StringHash(name) as number) || 0;
}

function 写局部(type: "unit" | "integer" | "real" | "string", name: string, value: any): void {
  const ydloc = 获取YDLOC();
  if (!ydloc) return;
  const page = 取GSIndex();
  const key = 字符串哈希(name);
  switch (type) {
    case "unit":
      jass.SaveUnitHandle(ydloc, page, key, value);
      return;
    case "integer":
      jass.SaveInteger(ydloc, page, key, value);
      return;
    case "real":
      jass.SaveReal(ydloc, page, key, value);
      return;
    case "string":
      jass.SaveStr(ydloc, page, key, value);
      return;
  }
}

function 读局部(type: "unit" | "integer" | "real" | "string", name: string): any {
  const ydloc = 获取YDLOC();
  if (!ydloc) return null;
  const page = 取GLIndex();
  const key = 字符串哈希(name);
  switch (type) {
    case "unit":
      return jass.LoadUnitHandle(ydloc, page, key);
    case "integer":
      return jass.LoadInteger(ydloc, page, key);
    case "real":
      return jass.LoadReal(ydloc, page, key);
    case "string":
      return jass.LoadStr(ydloc, page, key);
  }
}

function 获取YDLOC(): any {
  const g = globalThis as any;
  return g.YDLOC
    ?? jglobals.YDLOC
    ?? g.YDHASH_HANDLE
    ?? jglobals.YDHASH_HANDLE
    ?? g.YDHT
    ?? jglobals.YDHT
    ?? g.udg_YDHASH_HANDLE
    ?? jglobals.udg_YDHASH_HANDLE
    ?? g.udg_YDHT
    ?? jglobals.udg_YDHT
    ?? null;
}

function 取GSIndex(): number {
  const g = globalThis as any;
  const value = g.G_SIndex ?? jglobals.G_SIndex;
  return typeof value === "number" ? value : 0;
}

function 设GSIndex(v: number): void {
  (globalThis as any).G_SIndex = v;
  (jglobals as any).G_SIndex = v;
}

function 取GLIndex(): number {
  const g = globalThis as any;
  const value = g.G_LIndex ?? jglobals.G_LIndex;
  return typeof value === "number" ? value : 0;
}

function 设GLIndex(v: number): void {
  (globalThis as any).G_LIndex = v;
  (jglobals as any).G_LIndex = v;
}

function 触发STES事件(name: string): void {
  if (typeof STES_Fire_Global !== "function") return;
  STES_Fire_Global(null, name);
}

function 四位码(raw: string): number {
  return (
    raw.charCodeAt(0) * 16777216
    + raw.charCodeAt(1) * 65536
    + raw.charCodeAt(2) * 256
    + raw.charCodeAt(3)
  );
}

type VoidCallback = () => void;

export interface 召唤物参数 {
  主人单位: any;
  单位类型?: string | number;
  召唤物单位?: any;

  X: number;
  Y: number;
  朝向?: number;
  持续时间?: number;

  模型文件?: string;
  飞行高度?: number;
  生命值?: number;
  生命回复?: number;
  攻击力?: number;
  攻击间隔?: number;
  护甲?: number;
  缩放?: number;
}

const SUMMON_STES_EVENT = "OnSummonEvent";
const LOCAL_PAGE_BASE = 0x53000000;
let nextLocalPageSeed = 0;

function 归一化单位类型(unitType: string | number | undefined): number {
  if (typeof unitType === "number") return unitType;
  if (typeof unitType === "string" && unitType.length === 4) {
    return 四位码(unitType);
  }
  return 0;
}

function 分配临时父页(): number {
  nextLocalPageSeed += 1;
  if (nextLocalPageSeed >= 0x00ffffff) {
    nextLocalPageSeed = 1;
  }
  return LOCAL_PAGE_BASE + nextLocalPageSeed;
}

function 写入召唤参数(参数: 召唤物参数): void {
  写局部("unit", "Master", 参数.主人单位);

  if (参数.召唤物单位 != null && 参数.召唤物单位 !== 0) {
    写局部("unit", "Summon", 参数.召唤物单位);
  }

  const 单位类型 = 归一化单位类型(参数.单位类型);
  if (单位类型 !== 0) {
    写局部("integer", "unitType", 单位类型);
  }

  写局部("real", "x", 参数.X);
  写局部("real", "y", 参数.Y);
  写局部("real", "facing", 参数.朝向 ?? 0.0);
  写局部("real", "time", 参数.持续时间 ?? 0.0);
  写局部("string", "ModelFileID", 参数.模型文件 ?? "");

  if (参数.飞行高度 != null) {
    写局部("real", "moveHeight", 参数.飞行高度);
  }
  if (参数.生命值 != null) {
    写局部("real", "HP", 参数.生命值);
  }
  if (参数.生命回复 != null) {
    写局部("real", "regenHP", 参数.生命回复);
  }
  if (参数.攻击力 != null) {
    写局部("real", "AttackPower", 参数.攻击力);
  }
  if (参数.攻击间隔 != null) {
    // 旧 JASS 模板这里存在 `MoveHeight` / `atkCd` 的大小写/命名混用，统一双写兼容。
    写局部("real", "MoveHeight", 参数.攻击间隔);
    写局部("real", "atkCd", 参数.攻击间隔);
  }
  if (参数.护甲 != null) {
    写局部("real", "def", 参数.护甲);
  }
  if (参数.缩放 != null) {
    写局部("real", "size", 参数.缩放);
  }
}

export function 创建召唤物并套用JASS模板(参数: 召唤物参数): any {
  if (参数.主人单位 == null || 参数.主人单位 === 0) return null;
  if ((参数.召唤物单位 == null || 参数.召唤物单位 === 0) && 归一化单位类型(参数.单位类型) === 0) {
    return null;
  }

  const YDLOC = 获取YDLOC();
  if (!YDLOC) return null;

  const prevSIndex = 取GSIndex();
  const prevLIndex = 取GLIndex();
  const parentPage = 分配临时父页();

  设GSIndex(parentPage);
  设GLIndex(parentPage);
  写入召唤参数(参数);
  触发STES事件(SUMMON_STES_EVENT);

  const 召唤物 = 读局部("unit", "Summon");

  jass.FlushChildHashtable(YDLOC, parentPage);
  设GSIndex(prevSIndex);
  设GLIndex(prevLIndex);

  return 召唤物 != null && 召唤物 !== 0 ? 召唤物 : null;
}

export function 快捷创建召唤物(
  主人单位: any,
  单位类型: string | number,
  X: number,
  Y: number,
  持续时间: number,
  额外参数?: Omit<召唤物参数, "主人单位" | "单位类型" | "X" | "Y" | "持续时间">
): any {
  return 创建召唤物并套用JASS模板({
    主人单位,
    单位类型,
    X,
    Y,
    持续时间,
    ...额外参数,
  });
}

export {};

File: src/系统/03．技能系统/00．技能模板+函数/02．通用函数/05．召唤物.lua
local ____lualib = require("lualib_bundle")
local __TS__ObjectAssign = ____lualib.__TS__ObjectAssign
local ____exports = {}
local _____83B7_53D6YDLOC, _____53D6GSIndex, _____53D6GLIndex, jglobals
function _____83B7_53D6YDLOC()
    local g = _G
    local ____g_YDLOC_0 = g.YDLOC
    if ____g_YDLOC_0 == nil then
        ____g_YDLOC_0 = jglobals.YDLOC
    end
    local ____g_YDLOC_0_1 = ____g_YDLOC_0
    if ____g_YDLOC_0_1 == nil then
        ____g_YDLOC_0_1 = g.YDHASH_HANDLE
    end
    local ____g_YDLOC_0_1_2 = ____g_YDLOC_0_1
    if ____g_YDLOC_0_1_2 == nil then
        ____g_YDLOC_0_1_2 = jglobals.YDHASH_HANDLE
    end
    local ____g_YDLOC_0_1_2_3 = ____g_YDLOC_0_1_2
    if ____g_YDLOC_0_1_2_3 == nil then
        ____g_YDLOC_0_1_2_3 = g.YDHT
    end
    local ____g_YDLOC_0_1_2_3_4 = ____g_YDLOC_0_1_2_3
    if ____g_YDLOC_0_1_2_3_4 == nil then
        ____g_YDLOC_0_1_2_3_4 = jglobals.YDHT
    end
    local ____g_YDLOC_0_1_2_3_4_5 = ____g_YDLOC_0_1_2_3_4
    if ____g_YDLOC_0_1_2_3_4_5 == nil then
        ____g_YDLOC_0_1_2_3_4_5 = g.udg_YDHASH_HANDLE
    end
    local ____g_YDLOC_0_1_2_3_4_5_6 = ____g_YDLOC_0_1_2_3_4_5
    if ____g_YDLOC_0_1_2_3_4_5_6 == nil then
        ____g_YDLOC_0_1_2_3_4_5_6 = jglobals.udg_YDHASH_HANDLE
    end
    local ____g_YDLOC_0_1_2_3_4_5_6_7 = ____g_YDLOC_0_1_2_3_4_5_6
    if ____g_YDLOC_0_1_2_3_4_5_6_7 == nil then
        ____g_YDLOC_0_1_2_3_4_5_6_7 = g.udg_YDHT
    end
    local ____g_YDLOC_0_1_2_3_4_5_6_7_8 = ____g_YDLOC_0_1_2_3_4_5_6_7
    if ____g_YDLOC_0_1_2_3_4_5_6_7_8 == nil then
        ____g_YDLOC_0_1_2_3_4_5_6_7_8 = jglobals.udg_YDHT
    end
    local ____g_YDLOC_0_1_2_3_4_5_6_7_8_9 = ____g_YDLOC_0_1_2_3_4_5_6_7_8
    if ____g_YDLOC_0_1_2_3_4_5_6_7_8_9 == nil then
        ____g_YDLOC_0_1_2_3_4_5_6_7_8_9 = nil
    end
    return ____g_YDLOC_0_1_2_3_4_5_6_7_8_9
end
function _____53D6GSIndex()
    local g = _G
    local ____g_G_SIndex_10 = g.G_SIndex
    if ____g_G_SIndex_10 == nil then
        ____g_G_SIndex_10 = jglobals.G_SIndex
    end
    local value = ____g_G_SIndex_10
    return type(value) == "number" and value or 0
end
function _____53D6GLIndex()
    local g = _G
    local ____g_G_LIndex_11 = g.G_LIndex
    if ____g_G_LIndex_11 == nil then
        ____g_G_LIndex_11 = jglobals.G_LIndex
    end
    local value = ____g_G_LIndex_11
    return type(value) == "number" and value or 0
end
--- 通用函数 - 召唤物快捷模板
-- 
-- 说明：
-- - 配套底层 JASS 源文件：
--   `JASS/jass复制粘贴/召唤物.j`
-- - 本文件不重写 JASS 召唤逻辑，而是按旧模板要求写入 YDLocal 参数后触发 STES 事件 `OnSummonEvent`
-- - 适合技能侧快速创建/配置召唤物，并继续复用 JASS 端现有字段语义
local jass = require("jass.common")
jglobals = require("jass.globals")
local STES_Fire_Global = _G.STES_Fire
local function _____5B57_7B26_4E32_54C8_5E0C(name)
    return jass.StringHash(name) or 0
end
local function _____5199_5C40_90E8(____type, name, value)
    local ydloc = _____83B7_53D6YDLOC()
    if not ydloc then
        return
    end
    local page = _____53D6GSIndex()
    local key = _____5B57_7B26_4E32_54C8_5E0C(name)
    repeat
        local ____switch5 = ____type
        local ____cond5 = ____switch5 == "unit"
        if ____cond5 then
            jass.SaveUnitHandle(ydloc, page, key, value)
            return
        end
        ____cond5 = ____cond5 or ____switch5 == "integer"
        if ____cond5 then
            jass.SaveInteger(ydloc, page, key, value)
            return
        end
        ____cond5 = ____cond5 or ____switch5 == "real"
        if ____cond5 then
            jass.SaveReal(ydloc, page, key, value)
            return
        end
        ____cond5 = ____cond5 or ____switch5 == "string"
        if ____cond5 then
            jass.SaveStr(ydloc, page, key, value)
            return
        end
    until true
end
local function _____8BFB_5C40_90E8(____type, name)
    local ydloc = _____83B7_53D6YDLOC()
    if not ydloc then
        return nil
    end
    local page = _____53D6GLIndex()
    local key = _____5B57_7B26_4E32_54C8_5E0C(name)
    repeat
        local ____switch8 = ____type
        local ____cond8 = ____switch8 == "unit"
        if ____cond8 then
            return jass.LoadUnitHandle(ydloc, page, key)
        end
        ____cond8 = ____cond8 or ____switch8 == "integer"
        if ____cond8 then
            return jass.LoadInteger(ydloc, page, key)
        end
        ____cond8 = ____cond8 or ____switch8 == "real"
        if ____cond8 then
            return jass.LoadReal(ydloc, page, key)
        end
        ____cond8 = ____cond8 or ____switch8 == "string"
        if ____cond8 then
            return jass.LoadStr(ydloc, page, key)
        end
    until true
end
local function _____8BBEGSIndex(v)
    _G.G_SIndex = v
    jglobals.G_SIndex = v
end
local function _____8BBEGLIndex(v)
    _G.G_LIndex = v
    jglobals.G_LIndex = v
end
local function _____89E6_53D1STES_4E8B_4EF6(name)
    if type(STES_Fire_Global) ~= "function" then
        return
    end
    STES_Fire_Global(nil, name)
end
local function _____56DB_4F4D_7801(raw)
    return (string.byte(raw, 1) or 0 / 0) * 16777216 + (string.byte(raw, 2) or 0 / 0) * 65536 + (string.byte(raw, 3) or 0 / 0) * 256 + (string.byte(raw, 4) or 0 / 0)
end
local SUMMON_STES_EVENT = "OnSummonEvent"
local LOCAL_PAGE_BASE = 1392508928
local nextLocalPageSeed = 0
local function _____5F52_4E00_5316_5355_4F4D_7C7B_578B(unitType)
    if type(unitType) == "number" then
        return unitType
    end
    if type(unitType) == "string" and #unitType == 4 then
        return _____56DB_4F4D_7801(unitType)
    end
    return 0
end
local function _____5206_914D_4E34_65F6_7236_9875()
    nextLocalPageSeed = nextLocalPageSeed + 1
    if nextLocalPageSeed >= 16777215 then
        nextLocalPageSeed = 1
    end
    return LOCAL_PAGE_BASE + nextLocalPageSeed
end
local function _____5199_5165_53EC_5524_53C2_6570(_____53C2_6570)
    _____5199_5C40_90E8("unit", "Master", _____53C2_6570["主人单位"])
    if _____53C2_6570["召唤物单位"] ~= nil and _____53C2_6570["召唤物单位"] ~= 0 then
        _____5199_5C40_90E8("unit", "Summon", _____53C2_6570["召唤物单位"])
    end
    local _____5355_4F4D_7C7B_578B = _____5F52_4E00_5316_5355_4F4D_7C7B_578B(_____53C2_6570["单位类型"])
    if _____5355_4F4D_7C7B_578B ~= 0 then
        _____5199_5C40_90E8("integer", "unitType", _____5355_4F4D_7C7B_578B)
    end
    _____5199_5C40_90E8("real", "x", _____53C2_6570.X)
    _____5199_5C40_90E8("real", "y", _____53C2_6570.Y)
    _____5199_5C40_90E8("real", "facing", _____53C2_6570["朝向"] or 0)
    _____5199_5C40_90E8("real", "time", _____53C2_6570["持续时间"] or 0)
    _____5199_5C40_90E8("string", "ModelFileID", _____53C2_6570["模型文件"] or "")
    if _____53C2_6570["飞行高度"] ~= nil then
        _____5199_5C40_90E8("real", "moveHeight", _____53C2_6570["飞行高度"])
    end
    if _____53C2_6570["生命值"] ~= nil then
        _____5199_5C40_90E8("real", "HP", _____53C2_6570["生命值"])
    end
    if _____53C2_6570["生命回复"] ~= nil then
        _____5199_5C40_90E8("real", "regenHP", _____53C2_6570["生命回复"])
    end
    if _____53C2_6570["攻击力"] ~= nil then
        _____5199_5C40_90E8("real", "AttackPower", _____53C2_6570["攻击力"])
    end
    if _____53C2_6570["攻击间隔"] ~= nil then
        _____5199_5C40_90E8("real", "MoveHeight", _____53C2_6570["攻击间隔"])
        _____5199_5C40_90E8("real", "atkCd", _____53C2_6570["攻击间隔"])
    end
    if _____53C2_6570["护甲"] ~= nil then
        _____5199_5C40_90E8("real", "def", _____53C2_6570["护甲"])
    end
    if _____53C2_6570["缩放"] ~= nil then
        _____5199_5C40_90E8("real", "size", _____53C2_6570["缩放"])
    end
end
____exports["创建召唤物并套用JASS模板"] = function(_____53C2_6570)
    if _____53C2_6570["主人单位"] == nil or _____53C2_6570["主人单位"] == 0 then
        return nil
    end
    if (_____53C2_6570["召唤物单位"] == nil or _____53C2_6570["召唤物单位"] == 0) and _____5F52_4E00_5316_5355_4F4D_7C7B_578B(_____53C2_6570["单位类型"]) == 0 then
        return nil
    end
    local YDLOC = _____83B7_53D6YDLOC()
    if not YDLOC then
        return nil
    end
    local prevSIndex = _____53D6GSIndex()
    local prevLIndex = _____53D6GLIndex()
    local parentPage = _____5206_914D_4E34_65F6_7236_9875()
    _____8BBEGSIndex(parentPage)
    _____8BBEGLIndex(parentPage)
    _____5199_5165_53EC_5524_53C2_6570(_____53C2_6570)
    _____89E6_53D1STES_4E8B_4EF6(SUMMON_STES_EVENT)
    local _____53EC_5524_7269 = _____8BFB_5C40_90E8("unit", "Summon")
    jass.FlushChildHashtable(YDLOC, parentPage)
    _____8BBEGSIndex(prevSIndex)
    _____8BBEGLIndex(prevLIndex)
    local ____temp_12
    if _____53EC_5524_7269 ~= nil and _____53EC_5524_7269 ~= 0 then
        ____temp_12 = _____53EC_5524_7269
    else
        ____temp_12 = nil
    end
    return ____temp_12
end
____exports["快捷创建召唤物"] = function(_____4E3B_4EBA_5355_4F4D, _____5355_4F4D_7C7B_578B, X, Y, _____6301_7EED_65F6_95F4, _____989D_5916_53C2_6570)
    return ____exports["创建召唤物并套用JASS模板"](__TS__ObjectAssign({
        ["主人单位"] = _____4E3B_4EBA_5355_4F4D,
        ["单位类型"] = _____5355_4F4D_7C7B_578B,
        X = X,
        Y = Y,
        ["持续时间"] = _____6301_7EED_65F6_95F4
    }, _____989D_5916_53C2_6570))
end
return ____exports
